### ISSUE TITLE
Validator error message shows after blur not on change

### ISSUE THREAD

**Issue by regstar** [udos86/ng-dynamic-forms#914]
## I'm submitting a

<pre><code>
[ ] Bug / Regression
[ X ] Feature Request / Proposal
[ X ] Question
</code></pre>

## I'm using

<pre><code>
NG Dynamic Forms Version: `X.Y.Z`

[ ] Basic UI
[ ] Bootstrap UI  
[ ] Foundation UI
[ ] Ionic UI
[ ] Kendo UI
[ ] Material  
[ X ] NG Bootstrap
[ ] Prime NG
</code></pre>

## Description

Error messages with default and custom validators will shown up only if i blur out on my form control. When i type on a minLength textfield a smaller amount of text, the error message will only shown if i switch to another form element, or click out of my form. This state is not helpful for my formular. I need to display the error messages (if a error exist) immediately (onchange instead of blur out). Is there any method to do this?

**Comment by rzych** [comment]
Up. I think, this condition  
`return this.model.hasErrorMessages && this.control.touched && !this.hasFocus && this.isInvalid;`
should has optional hasFocus argument or something.

**Comment by regstar** [comment]
Yeah a optional flag would be nice!

**Comment by rernens** [comment]
@regstar 

hi,

a config option exists for this in DynamicFormControl
use updateOn: 'change' 

to accomplish what you are looking for.

**Comment by regstar** [comment]
Hey. Thx, i will test this.

**Comment by regstar** [comment]
Hello @rernens 

updateOn change is the standard value for this, but i added this on my options. But this doesn't work for me. It's the same state like the blur. Only if i click on another form control, the validation state will updating, and not on typing (on change).

**Comment by rernens** [comment]
@regstar 

Hi, which version of angular and ng-dynamic-forms are you using ?

**Comment by regstar** [comment]
@rernens

Angular 6.0.9
ng-dynamic-forms 6.0.7

**Comment by rernens** [comment]
@regstar

will check with these versions ASAP and get back to you

**Comment by rernens** [comment]
@regstar 

hi, 

I have created an environment with your versions as well as higher and I am experimenting the same behavior as you, while validators are well triggered on change when updateOn si set to change, messages are not displayed until blur.

Will try to identify why.

**Comment by rernens** [comment]
@regstar 

Hi again, 

Further investigation on this shows that this is related to change in angular material way of displaying messages on form elements . See below excerpt from Angular Material documentation

> Changing when error messages are shown
> 
> The <mat-form-field> allows you to associate error messages with your <select> or <mat-select>. By default, these error messages are shown when the control is invalid and either the user has interacted with (touched) the element or the parent form has been submitted. If you wish to override this behavior (e.g. to show the error as soon as the invalid control is dirty or when a parent form group is invalid), you can use the errorStateMatcher property of the <mat-select>. The property takes an instance of an ErrorStateMatcher object. An ErrorStateMatcher must implement a single method isErrorState which takes the FormControl for this <mat-select> as well as the parent form and returns a boolean indicating whether errors should be shown. (true indicating that they should be shown, and false indicating that they should not.)
> 
> A global error state matcher can be specified by setting the ErrorStateMatcher provider. This applies to all inputs. For convenience, ShowOnDirtyErrorStateMatcher is available in order to globally cause input errors to show when the input is dirty and invalid.
> 
```
@NgModule({
    providers: [
       {provide: ErrorStateMatcher, useClass: ShowOnDirtyErrorStateMatcher}
    ]
})
```

**Comment by Karamuto** [comment]
Hello @regstar, @rzych and @rernens ,

I like it that you are so much into this.

As @rzych has stated this part here is the trouble maker:

`return this.model.hasErrorMessages && this.control.touched && !this.hasFocus && this.isInvalid;`

This is a behaviour which can't be overrriden and is used by all libraries except material, as they have the ErrorStateMatcher's which was mentioned by @regstar .

I can only see a solution in providing something like the error state matchers in a global fashion for all libraries to have the opportunity to override it as one pleases.

Greetings

**Comment by regstar** [comment]
Hello @Karamuto 
Can you tell me if this topic is being worked on? Maybe even a time of completion?

**Comment by Karamuto** [comment]
Hello @regstar ,

I don't think someone is currently working on this issue. Maybe you want to try to implement it yourself?

**Comment by regstar** [comment]
@Karamuto 
Hey. I have added on my fork some changes to handle this problem. Is it possible that you have a look at the fork before I commit it on this development branch?

I use the updateOn option from the form model to display the error message, like on blur or on change. The standard value will be on change. Maybe you have a better idea to handle this. Big thx!

https://github.com/regstar/ng-dynamic-forms/commit/50f6e7d0994bbc117928b17904a5b3365cd7bb12

**Comment by Karamuto** [comment]
Hi @regstar ,
now this is suprising half a year later :)

I took a short peek at you code. Maybe you want to add test cases to the unit tests to check if it does what you intend to do?

Greetings

**Comment by regstar** [comment]
@Karamuto Yeah a lot of work at the moment ;). I will add unit tests for this. Thx for your help!

**Comment by regstar** [comment]
@Karamuto Pull request added for this: https://github.com/udos86/ng-dynamic-forms/pull/1040

**Comment by udos86** [comment]
fixed in 10.1.0

see new feature in CHANGELOG.md
